scripts/exporters/gltf-exporter.js

```diff
@@ -0,0 +1,446 @@
+                const opacity = mat.opacity;
+                const blendType = mat.blendType;
+                const cull = mat.cull;
+
+                const material = {};
+
+                if (name && name.length > 0) {
+                    material.name = name;
+                }
+
+                if (!diffuse.equals(pc.Color.WHITE) || opacity !== 1) {
+                    material.pbrMetallicRoughness = {};
+                    material.pbrMetallicRoughness.baseColorFactor = [diffuse.r, diffuse.g, diffuse.b, opacity];
+                }
+
+                if (!emissive.equals(pc.Color.BLACK)) {
+                    material.emissiveFactor = [emissive.r, emissive.g, emissive.b];
+                }
+
+                if (blendType === pc.BLEND_NORMAL) {
+                    material.alphaMode = "BLEND";
+                }
+
+                if (cull === pc.CULLFACE_NONE) {
+                    material.doubleSided = true;
+                }
+
+                json.materials.push(material);
+            });
+        }
+    }
+
+    writeNodes(resources, json) {
+        if (resources.entities.length > 0) {
+            json.nodes = [];
+
+            resources.entities.forEach(function (entity) {
+                const name = entity.name;
+                const t = entity.getLocalPosition();
+                const r = entity.getLocalRotation();
+                const s = entity.getLocalScale();
+
+                const node = {};
+
+                if (name && name.length > 0) {
+                    node.name = name;
+                }
+
+                if (!t.equals(pc.Vec3.ZERO)) {
+                    node.translation = [t.x, t.y, t.z];
+                }
+
+                if (!r.equals(pc.Quat.IDENTITY)) {
+                    node.rotation = [r.x, r.y, r.z, r.w];
+                }
+
+                if (!s.equals(pc.Vec3.ONE)) {
+                    node.scale = [s.x, s.y, s.z];
+                }
+
+                if (entity.camera && entity.camera.enabled) {
+                    node.camera = resources.cameras.indexOf(entity.camera);
+                }
+
+                if (entity.render && entity.render.enabled) {
+                    entity.render.meshInstances.forEach((meshInstance) => {
+                        node.mesh = resources.meshInstances.indexOf(meshInstance);
+                    });
+                }
+
+                if (entity.model && entity.model.enabled) {
+                    entity.model.meshInstances.forEach((meshInstance) => {
+                        node.mesh = resources.meshInstances.indexOf(meshInstance);
+                    });
+                }
+                if (entity.children.length > 0) {
+                    node.children = [];
+
+                    entity.children.forEach((child) => {
+                        node.children.push(resources.entities.indexOf(child));
+                    });
+                }
+
+                json.nodes.push(node);
+            });
+        }
+    }
+
```

Comment:
```suggestion
    assignMeshIDs(resources, json) {
        var meshInstances = resources.meshInstances;
        // Assign meshes to JSON nodes based on meshInstance.node
        meshInstances.forEach(meshInstance => {
            const entityIndex = resources.entities.indexOf(meshInstance.node);
            if (entityIndex != -1) {
                json.nodes[entityIndex].mesh = meshInstances.indexOf(meshInstance);
            } else {
                console.warn('GltfExporter#assignMeshIDs> meshInstance referring to unexported json node');
            }
        });
    }
```

Reply:
Since there is still only one node for each entity, this will result in each meshInstance in an entity overwriting `json.nodes[entityIndex].mesh`, right? Should we be checking for meshInstances in models/renders in `writeNodes` and creating a new json node for each?

Reply:
This will be addressed in follow up PRs

Reply:
this is addressed here now: https://github.com/playcanvas/engine/pull/4683